### DISCUSSION TITLE
how to use frida "send()" function?

### DISCUSSION THREAD

**Discussion by WimpyMistake** [vfsfitvnm/frida-il2cpp-bridge#121] [Q&A]
i want to send message from by ts script to my python script. But the `send` function seems not sending anything. But when I put the `send` outside `Il2Cpp.perform` it does works. Below is the bare minimal ts code that I used.

```typescript
import "frida-il2cpp-bridge";

send("outside") // this works

Il2Cpp.perform(() => {
    // code here
    send("inside"); // this does nothing

});
```

also including the python script for reference.

```python
#!/usr/bin/python

import frida
import sys

def read_msg(message, data):
   print(message)
   print(data)

session = frida.get_usb_device().attach("com.the.game")

with open("./playground/_.js") as f:
    script = session.create_script(f.read(), runtime="v8")

script.on("message", read_msg)
script.load()
sys.stdin.read()
```

**Comment by vfsfitvnm** [comment]
Hmm, I cannot reproduce this bug
![Screenshot_20211219_150715](https://user-images.githubusercontent.com/46219656/146677856-2757dd08-eeba-4fed-b6f8-17e7f2297e7e.png)

I guess the callback (what you pass to `Il2Cpp.perform`) is not called at all?

**Comment by WimpyMistake** [comment]
ok this is strange. I'm using 0.6.6 which is latest 1.

> I guess the callback (what you pass to Il2Cpp.perform) is not called at all?

Its does. I did a `console.log('in 2')` and works.

I'm using node 14. Will it have any effect or anything? Anyway, thanks for your prompt reply.

**Comment by WimpyMistake** [comment]
hmm i'm not sure what happened but here is my workaround. I called the python script twice. The second time it works and received message from js. As in

```shell
python my-script.py
```

then `ctrl+c`

then again execute the script
```shell
python my-script.py
```

while I'm confused but I'm content with the workaround 😆 . Thanks for your input and great library.

**Reply by notwhy** [reply]
when you start , first time it print "outside" then you can try reload method
```
script_instance.unload()
with codecs.open(r".\_.js" , 'r' , 'utf-8') as f:
    script_text = f.read()
script_instance = frida_session.create_script(script_text)
script_instance.on('message' , on_message)
script_instance.load()
FRIDA_RPC = script_instance.exports
```
after this it should print inside
it looks like your game not successful load libil2cpp.so on the first time.

**Comment by vfsfitvnm** [comment]
Oh, I'm glad you figured out a workaround! I honestly don't know what was going on.
